src/main/java/io/dockstore/provision/DOSPlugin.java

```diff
@@ -49,24 +52,65 @@ public void stop() {
     @Extension
     public static class DOSPreProvision implements PreProvisionInterface {
 
-        DOSPluginUtil dosPluginUtil = new DOSPluginUtil();
         static final Set<String> SCHEME = new HashSet<>(Lists.newArrayList("dos"));
+        static final String SCHEME_PREFERENCE = "scheme-preference";
+        static final int GET_SCHEME = 0;
+        private Map<String, String> config;
+
+        DOSPluginUtil dosPluginUtil = new DOSPluginUtil();
+        List<String> preferredSchemes = new ArrayList<>();
+
+        @Override
+        public void setConfiguration(Map<String, String> map) {
+            this.config = map;
+
+            // Parse the preferred schemes into a list by splitting on commas and extra spaces, and remove all empty strings from the list
+            if (this.config.containsKey(SCHEME_PREFERENCE)) {
+
+                // Arrays.asList returns a fixed-sized array, so wrapping it as an ArrayList object allows list modification
+                this.preferredSchemes = new ArrayList<>(Arrays.asList(this.config.get(SCHEME_PREFERENCE).trim().split(",\\s*")));
+                this.preferredSchemes.removeIf(e -> e.equals(""));
+            }
+        }
 
         public Set<String> schemesHandled() {
             return SCHEME;
         }
 
         public List<String> prepareDownload(String targetPath) {
             List<String> urlList = new ArrayList<>();
+
+            // Linked Hash Maps ensure insertion-ordered key-value pairs. URLs resolved by the plugin are
+            // inserted by scheme (key) and a list of that scheme's URL(s) (value)
+            Map<String, List<String>> urlMap = new LinkedHashMap<>();
+            String protocol = ":\\/\\/(.+)/";
+
             Optional<ImmutableTriple<String, String, String>> uri = dosPluginUtil.splitURI(targetPath);
 
             if (uri.isPresent() && schemesHandled().contains(uri.get().getLeft())) {
                 Optional<JSONObject> jsonObj = dosPluginUtil.getResponse(uri.get());
 
                 if(jsonObj.isPresent()) {
                     JSONArray urls = jsonObj.get().getJSONObject("data_object").getJSONArray("urls");
+                    // Place all URLs into map for fast lookup time. URLs with duplicate schemes are appended to existing entry
                     for (int i = 0; i < urls.length(); i++) {
-                        urlList.add(urls.getJSONObject(i).getString("url"));
+                        String url = urls.getJSONObject(i).getString("url");
+                        String scheme = url.split(protocol)[GET_SCHEME];
+                        if (urlMap.containsKey(scheme)) {
```

Comment:
Could probably be simplified with something like https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#merge-K-V-java.util.function.BiFunction-